gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java

```diff
@@ -46,9 +146,52 @@ public int code() {
   }
 
   /**
+   * Returns the reason that caused the exception.
+   */
+  public String reason() {
+    return reason;
+  }
+
+  /**
    * Returns {@code true} when it is safe to retry the operation that caused this exception.
    */
   public boolean retryable() {
     return retryable;
   }
+
+  /**
+   * Returns {@code true} when the operation that caused this exception had no side effects.
+   */
+  public boolean idempotent() {
+    return idempotent;
+  }
+
+  protected static String reason(GoogleJsonError error) {
+    if (error.getErrors() != null  && !error.getErrors().isEmpty()) {
+      return error.getErrors().get(0).getReason();
+    }
+    return null;
+  }
+
+  protected static Error error(GoogleJsonError error) {
+    return new Error(error.getCode(), reason(error));
+  }
+
+  protected static String message(IOException exception) {
+    if (exception instanceof GoogleJsonResponseException) {
+      return ((GoogleJsonResponseException) exception).getDetails().getMessage();
+    }
+    return exception.getMessage();
+  }
+
+  protected static BaseServiceException translateAndThrow(
+      RetryHelper.RetryHelperException ex) {
+    if (ex.getCause() instanceof BaseServiceException) {
+      throw (BaseServiceException) ex.getCause();
+    }
+    if (ex instanceof RetryHelper.RetryInterruptedException) {
+      RetryHelper.RetryInterruptedException.propagate();
+    }
+    return null;
```

Comment:
Similar to guava's [Thrwoables](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Throwables.html) previously we would both
throw and return so one could invoke the method as `throw translateAndThrow` if desired.

Now, when this method can return `null` I see no reason for the return value.
Also, maybe we should consider rename it to `propagateIfPossible` or `translateAndPropagateIfPossible`?

Another option could get a "default" exception to throw if could not propagate (which seems to match current usage pattern).